### DISCUSSION TITLE
multiple packages for multiple executables?

### DISCUSSION THREAD

**Discussion by peterbecich** [srid/haskell-flake#378] [Q&A] [answered]
My understanding of the example ~https://github.com/Platonic-Systems/purescript-multi-nix/tree/master~ https://github.com/srid/haskell-multi-nix is that it shows how to package multiple `*.cabal` files and one `cabal.project` with one `haskell-flake`.

For a project with only one `*.cabal` file, with:
```
executable foo
  ...
executable bar
```
I believe `haskell-flake` will produce a package for `foo`.

How can I make packages for both?  Thanks

**Comment by shivaraj-bh** [comment, marked as answer]
`nix build .#foo` works, `nix build .#bar` should fail. But both `nix run .#foo` and `nix run .#bar` should run.

## Why?

Flake outputs’ `packages` exported by `haskell-flake` comes from `package` output of the function (See [1](https://github.com/srid/haskell-flake/blob/725292998cdb695dc312325603a662afa97b9e86/nix/modules/project/outputs.nix#L121) and [2](https://github.com/srid/haskell-flake/blob/725292998cdb695dc312325603a662afa97b9e86/nix/modules/projects.nix#L44)): https://github.com/srid/haskell-flake/blob/725292998cdb695dc312325603a662afa97b9e86/nix/modules/project/outputs.nix#L93-L109

All the other `executables`, including `foo`, in your cabal file are stored in `exes` of `buildPackageInfo`, which are exported in flake `apps`:
https://github.com/srid/haskell-flake/blob/725292998cdb695dc312325603a662afa97b9e86/nix/modules/project/outputs.nix#L123-L126

@srid any reason why the `exes` are not part of Flake outputs’ `packages`?

**Reply by srid** [reply]
> @srid any reason why the `exes` are not part of Flake outputs’ `packages`?

@peterbecich What is the larger (original) use case? Is there a repro? What was expected? What is the actual case? Starting from that would be beneficial to address this.

**Reply by peterbecich** [reply]
@shivaraj-bh , @srid , you are both correct, thank you. I overlooked the `apps`.

A repro is https://github.com/haskell/hackage-server.
This has in a single `cabal` file:
```
executable hackage-server
...
executable hackage-mirror
...
executable hackage-build
...
executable hackage-import
...
```


Those are runnable in `haskell-flake`:
```
$ nix run .#hackage-server
no command given (try --help)

$ nix run .#hackage-mirror
hackage-mirror: Expected path to a config file.
See hackage-mirror --help for details and an example.

$ nix run .#hackage-build 
Usage: hackage-build init URL [auxiliary URLs] [options]
       hackage-build build [packages] [options]

```
(`nix run .#hackage-import` fails for some reason)

I was confused by the failure of `nix build ...` as you noted @shivaraj-bh , and also the absence of these executables in `nix flake show` `packages`. These executables are listed in `apps`:
```
$ nix flake show            
...
├───apps
...
│   └───x86_64-linux
│       ├───default: app
│       ├───hackage-build: app
│       ├───hackage-import: app
│       ├───hackage-mirror: app
│       ├───hackage-server: app
│       └───mirror-hackage-server: app
...
└───packages
     ....
    └───x86_64-linux
        ├───default: package 'hackage-server-0.5.1'
        └───hackage-server: package 'hackage-server-0.5.1'
```

**Comment by srid** [comment]
> https://github.com/Platonic-Systems/purescript-multi-nix/tree/master

This repo has no Haskell packages. Did you mean to link to ~~https://github.com/srid/haskell-flake~~ https://github.com/srid/haskell-multi-nix?

In any case, for the use-case mentioned above - I'd normally expect the main cabal package to have both executables, i.e., `$(nix build .#<name>)/bin/{foo,bar}` should both exist.

**Reply by peterbecich** [reply]
Whoops, wrong link. I meant https://github.com/srid/haskell-multi-nix  Responding above

**Comment by roberth** [comment]
`haskell-flake` wraps the Nixpkgs Haskell infrastructure, based on [cabal2nix](https://github.com/NixOS/cabal2nix), which produces one derivation per _package_.

[haskell.nix](https://github.com/input-output-hk/haskell.nix) implements a derivation per _component_ but is slow to evaluate unless workarounds are applied. Those workarounds could perhaps be facilitated by a flake-parts module, but I am not sufficiently familiar with that Haskell integration, so I was unable to validate my mediocre attempt to package it for flake-parts against a baseline UX.

I'm happy to use a conceptually simpler integration like this one instead, even if the build is less fine grained.
